### PR TITLE
[runtime] Make external executables use full path

### PR DESCRIPTION
**Pull request by alexanderkyte**
The issue of using the name "mono" without an absolute path
means that if the PATH is changed during scripted runs,
one can end up using the wrong mono.

### CODE REVIEW

File: mcs/build/platforms/darwin.make [hunk 1]
```diff
@@ -9,9 +9,9 @@ PLATFORM_RUNTIME = $(RUNTIME)
 PLATFORM_CORLIB = mscorlib.dll
 PLATFORM_TEST_HARNESS_EXCLUDES = NotOnMac,
 
-EXTERNAL_MCS = mcs
-EXTERNAL_MBAS = mbas
-EXTERNAL_RUNTIME = mono
+EXTERNAL_MCS := $(shell which mcs)
```
Comment (alexrp) [inline mcs/build/platforms/darwin.make]: Why the change from `=` to `:=`?

**Comment by alexrp** [comment]
Looks OK otherwise.

File: mcs/build/platforms/darwin.make (hunk 1 above)
Comment (alexanderkyte) [inline mcs/build/platforms/darwin.make]: Because if it's "=" then this is done lazily. If this were computed when another script had changed the PATH, then the external mono might end up pointing to something else. We'd have the same problem we have when using a relative path.

Comment (akoeplinger) [inline mcs/build/platforms/darwin.make]: Why not do it for the other options as well then?

**Comment by akoeplinger** [comment]
We're doing stuff like this in a bunch of places, so I'm curious where/how you ran into this?

**Comment by alexanderkyte** [comment]
I use the external mono to run ilasm.exe for mobile_static, and when doing a full build/test it ended up pointing to the tmpinst mono.

File: mcs/build/platforms/darwin.make (hunk 1 above)
Comment (evincarofautumn) [inline mcs/build/platforms/darwin.make]: In fact most of our Makefile variables could use `:=`, not `=`, and this would probably improve build performance. A separate PR is probably in order.

**Comment by alexanderkyte** [comment]
All variable assignments are now strict for the external commands

**Comment by alexanderkyte** [comment]
build

**Comment by alexanderkyte** [comment]
This is no longer necessary as ilasm has had it's api reduced on mobile_static so it works with in-tree mono still. The fix might still be useful though, as it prevents a class of bugs related to dirty environments.

**Comment by alexanderkyte** [comment]
build

**Comment by alexanderkyte** [comment]
How do people feel about this? Should I close it or will we merge it? It's not required now, so it's really a matter of preference.
